ocaml/tests/test_certificates.ml

```diff
@@ -0,0 +1,206 @@
+(** This module tests the PKI validation
+ *)
+open Certificates
+open Api_errors
+
+(* Initialize RNG for testing certificates *)
+let () = Nocrypto_entropy_unix.initialize ()
+
+let time_of_rfc3339 date =
+  match Ptime.of_rfc3339 date with
+  | Ok (time, _, _) -> time
+  | Error _ -> raise (Failure ("Date is in the wrong format: " ^ date))
+
+let valid_from = time_of_rfc3339 "2020-01-01T00:00:00+00:00"
+
+let valid_until = time_of_rfc3339 "2021-01-01T00:00:00+00:00"
+
+let host_name = X509.Distinguished_name.Relative_distinguished_name.singleton (DC "localhost")
+
+let load_test_data file = Stdext.Unixext.string_of_file @@ "test_data/certificates/" ^ file ^ ".pem"
+
+let valid_private_keys = ["pkey_rsa_2048"; "pkey_rsa_4096"]
+
+(* ( file_name, error_type, error_message list ) *)
+let invalid_private_keys =
+  [ "pkey_rsa_1024", server_certificate_key_rsa_length_not_supported,
+    [ "1024" ]
+  ; "pkey_rsa_8192", server_certificate_key_rsa_length_not_supported,
+    [ "8192" ]
+  ; "pkey_rsa_n3_2048", server_certificate_key_rsa_multi_not_supported, []
+  ; "pkey_ed25519", server_certificate_key_algorithm_not_supported,
+    [ "1.3.101.112" ]
+  ; "pkey_bogus", server_certificate_key_invalid, []
+  ]
+
+(* ( description, leaf_private_key, time_of_validation, signature_algorithm ) *)
+let valid_leaf_certificates =
+  [ "Valid, SHA256, matches key",
+    "pkey_rsa_2048",
+    "2020-02-01T00:00:00+00:00",
+    `SHA256 ]
+
+(* ( description, leaf_private_key, expected_private_key, time_of_validation,
+     signature_algorithm, error_type, error_message list ) *)
+let invalid_leaf_certificates =
+  [ "Not valid yet, SHA256, matching key",
+    "pkey_rsa_2048",
+    "pkey_rsa_2048",
+    "2019-01-01T00:00:00+00:00",
+    `SHA256,
+    server_certificate_not_valid_yet,
+    [ "2019-01-01T00:00:00-00:00"; "2020-01-01T00:00:00-00:00" ]
+  ; "Expired, SHA256, matching key",
+    "pkey_rsa_2048",
+    "pkey_rsa_2048",
+    "2022-01-01T00:00:00+00:00",
+    `SHA256,
+    server_certificate_expired,
+    [ "2022-01-01T00:00:00-00:00"; "2021-01-01T00:00:00-00:00" ]
+  ; "Valid, SHA256, keys do not match",
+    "pkey_rsa_2048",
+    "pkey_rsa_4096",
+    "2020-02-01T00:00:00+00:00",
+    `SHA256,
+    server_certificate_key_mismatch,
+    []
+  ; "Valid, SHA1, matching keys",
+    "pkey_rsa_2048",
+    "pkey_rsa_2048",
+    "2020-02-01T00:00:00+00:00",
+    `SHA1,
+    server_certificate_signature_not_supported,
+    []
+  ]
+
+(* ( certificate_name, leaf_private_key, time_of_validation, error_tye,
+     error_message list ) *)
+let corrupt_certificates =
+  [ "cert_bogus",
+    "pkey_rsa_2048",
+    "2020-02-01T00:00:00+00:00",
+    server_certificate_invalid,
+    []
+  ]
+
+let key_chain = List.map (fun _ -> `RSA (Nocrypto.Rsa.generate 2048)) (List.init 3 Fun.id)
```

Comment:
is this not equivalent to `List.init 3 (`RSA (Nocrypto.Rsa.generate 2048))`?

Reply:
I agree this can be made shorter.